File: analyzer.py
from typing import Tuple, Callable
from stringdata import get_data
from time import time


def main():
    data = get_data()  # Access the data set from the stringdata module

    result, runtime = time_of_algorithm(linear_search, data, "not_here")
    print(f'Result of Linear Search for "not_here": {result}')
    print(f"Time: {runtime}", end="\n\n")

    result, runtime = time_of_algorithm(binary_search, data, "not_here")
    print(f'Result of Binary Search for "not_here": {result}')
    print(f"Time: {runtime}", end="\n\n")

    result, runtime = time_of_algorithm(linear_search, data, "mzzzz")
    print(f'Result of Linear Search for "mzzzz": {result}')
    print(f"Time: {runtime}", end="\n\n")

    result, runtime = time_of_algorithm(binary_search, data, "mzzzz")
    print(f'Result of Binary Search for "mzzzz": {result}')
    print(f"Time: {runtime}", end="\n\n")

    result, runtime = time_of_algorithm(linear_search, data, "aaaaa")
    print(f'Result of Linear Search for "aaaaa": {result}')
    print(f"Time: {runtime}", end="\n\n")

    result, runtime = time_of_algorithm(binary_search, data, "aaaaa")
    print(f'Result of Binary Search for "aaaaa": {result}')
    print(f"Time: {runtime}", end="\n\n")


# Finds runtime of called algorithm and returns result of algorithm
def time_of_algorithm(
    search: Callable[[Tuple[str], str], int], data: Tuple[str], element: str
) -> Tuple[str, str]:
    start_time = time()
    result = search(data, element)
    end_time = time()
    runtime = end_time - start_time
    return str(result), str(runtime)


# O(logn)
def binary_search(container: Tuple[str], element: str) -> int:
    low = 0  # First index of container
    high = len(container) - 1  # Last index of container
    mid = midpoint(low, high)
    old_mid = -1

    while old_mid != mid:  # Break when we check the same spot
        if container[mid] == element:
            return mid
        elif element < container[mid]:  # Need to look earlier in container
            old_mid = mid
            high = mid - 1  # New high is everything before mid
            mid = midpoint(low, high)
        else:  # Need to look later in container
            old_mid = mid
            low = mid + 1  # New low is everything after mid
            mid = midpoint(low, high)

    return -1


# O(n)
def linear_search(container: Tuple[str], element: str) -> int:
    for i in range(0, len(container)):
        if container[i] == element:
            return i

    return -1


def midpoint(low: int, high: int) -> int:
    return (low + high) // 2


if __name__ == "__main__":
    main()

File: analyzer.cpp
#include "StringData.h"
#include <iostream>

#define MIDPOINT(a, b) ((a + b) / 2)

using namespace std;

int linearSearch(const vector<string> &container, const string &element);

int binarySearch(const vector<string> &container, const string &element);

vector<string>
timeOfAlgorithm(int (&search)(const vector<string> &, const string &),
                const vector<string> &data, const string &element);

int main() {
  const vector<string> data =
      getStringData(); // Access the data set from the stringdata module

  vector<string> result = timeOfAlgorithm(linearSearch, data, "not_here");
  cout << "Result of Linear Search for \"not_here\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;

  result = timeOfAlgorithm(binarySearch, data, "not_here");
  cout << "Result of Binary Search for \"not_here\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;

  result = timeOfAlgorithm(linearSearch, data, "mzzzz");
  cout << "Result of Linear Search for \"mzzzz\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;

  result = timeOfAlgorithm(binarySearch, data, "mzzzz");
  cout << "Result of Binary Search for \"mzzzz\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;

  result = timeOfAlgorithm(linearSearch, data, "aaaaa");
  cout << "Result of Linear Search for \"aaaaa\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;

  result = timeOfAlgorithm(binarySearch, data, "aaaaa");
  cout << "Result of Binary Search for \"aaaaa\": " << result[0] << endl;
  cout << "Time: " << result[1] << endl << endl;
}

// Finds runtime of called algorithm and returns result of algorithm
vector<string>
timeOfAlgorithm(int (&search)(const vector<string> &, const string &),
                const vector<string> &data, const string &element) {
  const long long startTime = systemTimeNanoseconds();
  const int result = search(data, element);
  const long long endTime = systemTimeNanoseconds();

  const long long runtime = endTime - startTime;
  return {to_string(result), to_string(runtime).append("ns")};
}

// O(log n)
int binarySearch(const vector<string> &container, const string &element) {
  int low = 0; // First index of container
  int high = static_cast<int>(container.size()) - 1; // Last index of container
  int mid = MIDPOINT(low, high);
  int oldMid = -1;

  while (oldMid != mid) {
    // Break when we check the same spot
    if (container[mid] == element) {
      return mid;
    } else if (element < container[mid]) {
      // Need to look earlier in container
      oldMid = mid;
      high = mid - 1; // New high is everything before mid
      mid = MIDPOINT(low, high);
    } else {
      // Need to look later in container
      oldMid = mid;
      low = mid + 1; // New low is everything after mid
      mid = MIDPOINT(low, high);
    }
  }
  return -1;
}

// O(n)
int linearSearch(const vector<string> &container, const string &element) {
  for (int i = 0; i < container.size(); i++) {
    if (container[i] == element) {
      return i;
    }
  }

  return -1;
}
